text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,217 @@
+
+We have a number of situations where one create defines an interface, and a different crate implements that interface with a canonical singleton:
+
+ - [`core::alloc::GlobalAlloc`](https://doc.rust-lang.org/nightly/core/alloc/trait.GlobalAlloc.html), chosen with [`#[global_allocator]`](https://doc.rust-lang.org/1.23.0/unstable-book/language-features/global-allocator.html)
+ - `panic_fmt` chosen with [`#[panic_implementation]`](https://github.com/rust-lang/rfcs/blob/master/text/2070-panic-implementation.md)
+ - The OOM hook, modified with [`std::alloc::{set,take}_alloc_error_hook`](https://doc.rust-lang.org/nightly/std/alloc/fn.set_alloc_error_hook.html)
+ - `std::hash:RandomState`, if https://github.com/rust-lang/rust/pull/51846 is merged, the `hashmap_random_keys` lang item.
+ - [`log::Log`](https://docs.rs/log/0.4.3/log/trait.Log.html) set with https://docs.rs/log/0.4.3/log/fn.set_logger.html
+
+Each of these is an instance of the same general pattern.
+But the solutions are all ad-hoc and distinct, adding burdening the user of Rust and rustc with needless extra information, and preventing more rapid prototyping.
+They also incur a run-time cost due to dynamism and indirection, which can lead to initializing bugs or bloat in space-constrained environments.
+
+`existential type` just covers the deferred definition of a type, and not the singleton value itself, but that is enough. For example, with global allocation:
+
+```rust
+// in `alloc`
+
+pub extern existential type Heap: Copy + Alloc + Default;
+
+struct Box<T, A: Alloc = Heap>;
+
+impl Box<T, A: Alloc> {
+    fn new_in(a: A) { .. }
+}
+
+impl Box<T, A: Alloc + Default = Heap> {
+    fn new() { Self::new_in(Default::default())    }
+}
+```
+
+```rust
+// in `jemalloc`
+
+struct JemallocHeap;
+
+impl Default for JemallocHeap {
+    fn default() { JemallocHeap }
+}
+
+impl Alloc for JemallocHeap {
+    fn alloc(stuff: Stuff) -> Thing {
+        ...
+    }
+}
+
+impl existential type alloc::Heap = JemallocHeap;
+```
+
+```rust
+// in crate making an rust-implemented local allocator global
+
+static GLOBALIZED_LOCAL_ALLOC = MyConcurrentLocalAlloc(..):
+
+strict MyConcurrentLocalAllocHeap;
+
+impl Default for MyConcurrentLocalAllocHeap {
+    fn default() { MyConcurrentLocalAllocHeap }
+}
+
+impl Alloc for MyConcurrentLocalAllocHeap {
+    fn alloc(stuff: Stuff) -> Thing {
+        GLOBALIZED_LOCAL_ALLOC.alloc(stuff)
+    }
+}
+
+impl existential type alloc::Heap = JemallocHeap;
+```
+
+By defining traits for each bit of deferred functionary, we can likewise cover each of the other use-cases.
+This frees the compiler and programmer to forget about the specific instances and just learn the general pattern.
+This is especially important for `log`, which isn't a sysroot crate and thus isn't known to the compiler at all at the moment.
+It would be very hard to justify special casing it with e.g. another attribute as the problem is solved today, when it needs none at the moment.
```

Comment:
"[..] special casing *it* [..]"

What is `it` referring to? Spell out what it is and remove *"it"*.